pkg/applicationserver/applicationserver.go

```diff
@@ -1352,6 +1412,26 @@ func (as *ApplicationServer) handleDownlinkNack(ctx context.Context, ids *ttnpb.
 			}
 
 			items := []*ttnpb.ApplicationDownlink{msg}
+
+			if msg.ConfirmedRetry == nil {
+				msg.ConfirmedRetry = &ttnpb.ApplicationDownlink_ConfirmedRetry{
+					Attempt: 0,
+				}
+			}
+			msg.ConfirmedRetry.Attempt++
```

Comment:
Though `msg` is a pointer, you're updating it after it's added to the `items` which is then passed to `attemptDownlinkQueueOp`.
Or am I understanding this wrongly?

Reply:
Yes, it's updated after being added to the array, but indeed, it is a pointer, so the value of `msg.ConfirmedRetry.Attempt` is updated before invoking `attemptDownlinkQueueOp`. It is also passed in the `registerDropDownlinks`.

Would you rather see the update before the addition to the array?

Reply:
Yes please. Though the end result is the same, in the current implementation the update is a side-effect rather than being intentional.